examples/SmoothedParticleHydrodynamics/src/model/functions.c

```diff
@@ -0,0 +1,233 @@
+
+/*
+ * FLAME GPU v 1.5.X for CUDA 9
+ * Copyright University of Sheffield.
+ * Original Author: Dr Paul Richmond (user contributions tracked on https://github.com/FLAMEGPU/FLAMEGPU)
+ * Contact: p.richmond@sheffield.ac.uk (http://www.paulrichmond.staff.shef.ac.uk)
+ *
+ * University of Sheffield retain all intellectual property and
+ * proprietary rights in and to this software and related documentation.
+ * Any use, reproduction, disclosure, or distribution of this software
+ * and related documentation without an express license agreement from
+ * University of Sheffield is strictly prohibited.
+ *
+ * For terms of licence agreement please attached licence or view licence
+ * on www.flamegpu.com website.
+ *
+ */
+
+/* 
+	Smoothed Particle Hydrodynamics Demo
+	Author: Matthew Leach
+	Contact: M.Leach@sheffield.ac.uk
+
+	Suggested things to try:
+	
+	Play with the viscosity to try fluids of different thickness - the simulation is
+	stable for viscosity values between ~0.1 - 30
+
+	Lower the fluid rest density to TODO see the 'FLAME' explode instead of fall
+
+	Use the python script in the iterations folder to change the initial arrangement of the fluid particles
+*/
+
+
+#ifndef _FLAMEGPU_FUNCTIONS
+#define _FLAMEGPU_FUNCTIONS
+
+#include <header.h>
+
+// Mathematic constants
+#define PI 3.141593f
+
+// Sim parameters
+#define TIMESTEP 0.001f
+#define SMOOTHING_LENGTH 0.040f
+#define MAX_VELOCITY 4.0f
+
+// Environemental parameters
+#define RESTITUTION_COEFFICIENT 0.9f
+#define HALF_BOUNDARY_WIDTH 0.5f
+
+// Fluid properties
+#define PARTICLE_MASS 0.001953125f
+#define PRESSURE_COEFFICIENT 0.9f
+#define FLUID_REST_DENSITY 1000.0f
+#define MINIMUM_PARTICLE_DENSITY 0.0f
+#define VISCOSITY 10.0f
+
+//// Math helpers
+inline __device__ float dot(glm::vec3 a, glm::vec3 b) {
```

Comment:
`__device__` could be replaced with `__FLAME_GPU_FUNC__` to maintain the lack of CUDA code in examples (it has the same semantic meaning, but masks `__device__` from modelers.